File: tip-calculator-app/app/controllers/calculator_controller.rb
class CalculatorController < ApplicationController
  # Initialize new calculator instance
  def index
    @tip_calculate = TipCalculate.new
  end
  
  # Handle AJAX calculation request
  def calculate
    @tip_calculate = TipCalculate.new(calculation_params)
    
    # Calculate derived values before saving
    calculate_amounts
    
    # Save calculation for admin dashboard and return result
    if @tip_calculate.save
      render json: {
        success: true,
        tip_amount: format_amount(@tip_calculate.tip_amount),
        total_amount: format_amount(@tip_calculate.total_amount)
      }
    else
      render json: {
        success: false,
        errors: @tip_calculate.errors.full_messages
      }, status: :unprocessable_entity
    end
  rescue StandardError => e
    render json: {
      success: false,
      errors: [e.message]
    }, status: :internal_server_error
  end

  private
  
  # Strong parameters for security
  def calculation_params
    params.require(:calculation).permit(:bill_amount, :tip_percentage, :number_of_people)
  end
  
  # Calculate tip and total amounts
  def calculate_amounts
    # Convert inputs to float for accurate calculations
    bill = @tip_calculate.bill_amount.to_f
    tip_rate = @tip_calculate.tip_percentage.to_f / 100
    people = @tip_calculate.number_of_people.to_f
    
    # Calculate total tip
    total_tip = bill * tip_rate
    
    # Calculate tip per person
    @tip_calculate.tip_amount = total_tip / people
    
    # Calculate total per person (bill share + tip share)
    @tip_calculate.total_amount = (bill / people) + @tip_calculate.tip_amount
    
    # Round to 2 decimal places
    @tip_calculate.tip_amount = @tip_calculate.tip_amount.round(2)
    @tip_calculate.total_amount = @tip_calculate.total_amount.round(2)
  end
  
  # Format amount to 2 decimal places as string
  def format_amount(amount)
    "%.2f" % amount
  end
end

File: tip-calculator-app/app/javascript/controllers/calculator_controller.js
import { Controller } from "@hotwired/stimulus";

// Connects to data-controller="calculator"

export default class extends Controller {
  // Define the HTML elements that will be used in the controller
  static targets = [
    "bill",
    "tipBtn",
    "people",
    "customTip",
    "tipAmount",
    "total",
    "resetBtn",
    "errorMessage",
  ];

  connect() {
    this.reset();
    console.log("Calculator controller connected");
  }

// Event listeners for user interactions
  calculate() {


    // Get the values from the HTML elements
    const bill = parseFloat(this.billTarget.value) || 0;
    const tipPercent = this.getSelectedTipPercentage();
    const numberOfPeople = parseInt(this.peopleTarget.value) || 0;

    // Enable reset button if any values are entered
    this.resetBtnTarget.disabled = !(
      bill > 0 ||
      tipPercent > 0 ||
      numberOfPeople > 0
    );

    // Only proceed with calculation if we have valid inputs
    if (bill > 0 && numberOfPeople > 0) {
      try {


        // Client-side calculation for immediate feedback
        const tipRate = tipPercent / 100;
        const totalTip = bill * tipRate;
        const tipPerPerson = totalTip / numberOfPeople;
        const totalPerPerson = bill / numberOfPeople + tipPerPerson;

        // Update the UI immediately
        this.tipAmountTarget.textContent = `$${tipPerPerson.toFixed(2)}`;
        this.totalTarget.textContent = `$${totalPerPerson.toFixed(2)}`;
        this.hideError();

        // Send to server for logging and saving to the database
        this.sendToServer(bill, tipPercent, numberOfPeople);
      } catch (error) {
        console.error("Calculation error:", error);
        this.showError(error.message || "Error calculating values");
      }
    } else {
      // Reset display if inputs are invalid
      this.tipAmountTarget.textContent = "$0.00";
      this.totalTarget.textContent = "$0.00";
    }
  }

  // Send calculation to server (for logging/admin purposes)
  async sendToServer(bill, tipPercent, numberOfPeople) {
    try {
      const response = await fetch("/calculate", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json",
          "X-CSRF-Token": document.querySelector('meta[name="csrf-token"]')
            ?.content,
        },
        body: JSON.stringify({
          calculation: {
            bill_amount: bill,
            tip_percentage: tipPercent,
            number_of_people: numberOfPeople,
          },
        }),
      });

      if (!response.ok) {
        const errorData = await response.json();
        console.error("Server validation error:", errorData.errors);
        if (errorData.errors && errorData.errors.length > 0) {
          this.showError(errorData.errors.join(", "));
        }
      }
    } catch (error) {
      console.error("Server communication error:", error);

    }
  }

  // Gets the currently selected tip percentage (from buttons or custom input)
  getSelectedTipPercentage() {


    // Check if any tip button is active
    const activeButton = this.tipBtnTargets.find((btn) =>
      btn.classList.contains("active")
    );

    if (activeButton) {

      // If a button is active, return its percentage
      return parseFloat(activeButton.textContent);
    } else if (this.customTipTarget.value) {

      // Use custom tip input if provided
      return parseFloat(this.customTipTarget.value);
    }

    return 0;
  }

  // Handles tip button selection
  selectTip(event) {

    // Remove active class from all buttons
    this.tipBtnTargets.forEach((btn) => btn.classList.remove("active"));

    // Make clicked button active
    event.currentTarget.classList.add("active");

    // Clear custom tip input
    this.customTipTarget.value = "";

    // Trigger calculation immediately
    this.calculate();
  }

  // Handle custom tip input
  customTipInput() {
    
    // When typing in custom tip, deselect any active buttons
    this.tipBtnTargets.forEach((btn) => btn.classList.remove("active"));

    // Trigger calculation immediately
    this.calculate();
  }

  // Resets the calculator to initial state
  reset() {
    this.billTarget.value = "";
    this.peopleTarget.value = "";
    this.customTipTarget.value = "";
    this.tipBtnTargets.forEach((btn) => btn.classList.remove("active"));
    this.tipAmountTarget.textContent = "$0.00";
    this.totalTarget.textContent = "$0.00";
    this.resetBtnTarget.disabled = true;
    this.hideError();
  }

  // Shows error message
  showError(message) {
    if (this.hasErrorMessageTarget) {
      this.errorMessageTarget.textContent = message;
      this.errorMessageTarget.classList.remove("hidden");
    } else {
      console.error("Error:", message);
    }
  }

  // Hides error message
  hideError() {
    if (this.hasErrorMessageTarget) {
      this.errorMessageTarget.classList.add("hidden");
    }
  }
}